python/tests/test_zarr.py

```diff
@@ -10,58 +10,129 @@
+def test_direct_store_access(store, xp):
     """Test accessing the GDS Store directly"""
 
-    module = pytest.importorskip(array_type)
-    a = module.arange(5, dtype="u1")
+    a = xp.arange(5, dtype="u1")
     store["a"] = a
     b = store["a"]
 
-    # Notice, GDSStore always returns a cupy array
-    assert type(b) is cupy.ndarray
-    cupy.testing.assert_array_equal(a, b)
+    # Notice, when not using getitems(), GDSStore returns bytes always
+    assert isinstance(b, bytes)
+    a.data == b
 
 
-def test_array(store):
-    """Test Zarr array"""
+@pytest.mark.parametrize("xp_write", ["numpy", "cupy"])
+@pytest.mark.parametrize("xp_read_a", ["numpy", "cupy"])
+@pytest.mark.parametrize("xp_read_b", ["numpy", "cupy"])
+def test_direct_store_access_getitems(store, xp_write, xp_read_a, xp_read_b):
+    """Test accessing the GDS Store directly using getitems()"""
 
-    a = cupy.arange(100)
-    z = zarr.array(
-        a, chunks=10, compressor=None, store=store, meta_array=cupy.empty(())
+    xp_read_a = pytest.importorskip(xp_read_a)
+    xp_read_b = pytest.importorskip(xp_read_b)
+    xp_write = pytest.importorskip(xp_write)
+    a = xp_write.arange(5, dtype="u1")
+    b = a * 2
+    store["a"] = a
+    store["b"] = b
+
+    res = store.getitems(
+        keys=["a", "b"],
+        contexts={
+            "a": {"meta_array": xp_read_a.empty(())},
+            "b": {"meta_array": xp_read_b.empty(())},
+        },
     )
+    assert isinstance(res["a"], xp_read_a.ndarray)
+    assert isinstance(res["b"], xp_read_b.ndarray)
+    cupy.testing.assert_array_equal(res["a"], a)
+    cupy.testing.assert_array_equal(res["b"], b)
+
+
+def test_array(store, xp):
+    """Test Zarr array"""
+
+    a = xp.arange(100)
+    z = zarr.array(a, chunks=10, compressor=None, store=store, meta_array=xp.empty(()))
+    assert isinstance(z.meta_array, type(a))
     assert a.shape == z.shape
     assert a.dtype == z.dtype
     assert isinstance(a, type(z[:]))
     cupy.testing.assert_array_equal(a, z[:])
 
 
-def test_group(store):
+def test_group(store, xp):
     """Test Zarr group"""
 
-    g = zarr.open_group(store, meta_array=cupy.empty(()))
+    g = zarr.open_group(store, meta_array=xp.empty(()))
     g.ones("data", shape=(10, 11), dtype=int, compressor=None)
     a = g["data"]
     assert a.shape == (10, 11)
     assert a.dtype == int
     assert isinstance(a, zarr.Array)
-    assert isinstance(a[:], cupy.ndarray)
+    assert isinstance(a.meta_array, type(xp.empty(())))
+    assert isinstance(a[:], xp.ndarray)
     assert (a[:] == 1).all()
 
 
+def test_open_array(store, xp):
+    """Test Zarr's open_array()"""
+
+    a = xp.arange(10)
+    z = zarr.open_array(
+        store,
+        shape=a.shape,
+        dtype=a.dtype,
+        chunks=(10,),
+        compressor=None,
+        meta_array=xp.empty(()),
+    )
+    z[:] = a
+    assert a.shape == z.shape
+    assert a.dtype == z.dtype
+    assert isinstance(a, type(z[:]))
```

Comment:
Why do you need to slice `z` here? Ah, it's a zarr array so you must slice to get the backing store?

Reply:
yes